R/plumber.R

```diff
@@ -871,43 +865,136 @@ plumber <- R6Class(
+      stopifnot(length(ui) == 1)
+      stopifnot(is.logical(ui) || is.character(ui))
+      if (isTRUE(ui)) {
+        ui <- "swagger"
+      }
+      if (is.character(ui) && is_ui_available(ui)) {
+        enabled <- TRUE
+      } else {
+        enabled <- FALSE
+        ui <- "__not_enabled__"
+      }
+      private$ui_info = list(
+        enabled = enabled,
+        ui = ui,
+        args = list(...)
+      )
+    },
+    #' @description Set UI callback to notify where the API is located.
+    #'
+    #' When set, it will be called with a character string corresponding
+    #' to the API UI url. This allows RStudio to open `swagger` UI when a
+    #' Plumber router [pr_run()] method is executed using default `plumber.ui.callback` option.
+    #'
+    #' See also: [pr_set_ui_callback()]
+    #' @param callback a callback function for taking action on UI url. (Also accepts `NULL` values to disable the `callback`.)
+    set_ui_callback = function(
+      callback = getOption('plumber.ui.callback', getOption('plumber.swagger.url', NULL))
+    ) {
+      # Use callback when defined
+      if (!length(callback) || !is.function(callback)) {
+        callback = as.null
+      }
+      private$ui_callback = callback
+    },
+    #' @description Set debug value to include error messages
+    #' @param debug `TRUE` provides more insight into your API errors.
+    #' @seealso [pr_set_debug()]
+    set_debug = function(debug = interactive()) {
+      stopifnot(length(debug) == 1)
+      private$debug = isTRUE(debug)
+    },
     #' @description Add a filter to plumber router
     #' @param name a character string. Name of filter
     #' @param expr an expr that resolve to a filter function or a filter function
     #' @param serializer a serializer function
+    #' @seealso [pr_filter()]
     filter = function(name, expr, serializer){
       filter <- PlumberFilter$new(name, expr, private$envir, serializer)
       private$addFilterInternal(filter)
     },
+    #' @description
+    #' Add a function to customize what is returned in `$get_api_spec()`.
+    #'
+    #' Note, the returned value will be sent through [serializer_unboxed_json()] which will turn all length 1 vectors into atomic values.
+    #' To force a vector to serialize to an array of size 1, be sure to call [as.list()] on your value. `list()` objects are always serialized to an array value.
+    #'
+    #' See also: [pr_set_api_spec()]
+    #' @param api This can be
+    #'   * an OpenAPI Specification formatted list object
+    #'   * a function that accepts the OpenAPI Specification autogenerated by `plumber` and returns a OpenAPI Specification formatted list object.
+    #'
+    #'  The value returned will not be validated for OAS compatibility.
+    set_api_spec = function(api = NULL) {
+      api_fun <-
+        if (is.null(api)) {
+          identity
+        } else {
+          if (!is.function(api)) {
+            # function to return the api object
+            function(...) {
```

Comment:
Let's be a bit clearer this should be single arg

```suggestion
            function(x) {
```